aea/agent.py

```diff
@@ -54,12 +54,70 @@ def is_stopped(self) -> bool:
         return self._is_stopped
 
 
+class Identity:
+    """
+    An identity are the public elements identifying an agent.
+
+    It can include:
+    - the agent name
+    - the public half of private / public key pairs
+    - the addresses
+    """
+
+    def __init__(self, name: str):
+        """
+        Instantiate the agent.
+
+        :param name: the name of the agent.
+        """
+        self._name = name
+        self._public_keys = None  # type: Optional[Dict[str, str]]
+        self._addresses = None  # type: Optional[Dict[str, str]]
+
+    @property
+    def name(self) -> str:
+        """Get the agent name."""
+        return self._name
+
+    @property
+    def public_keys(self) -> Dict[str, str]:
+        """Get the public_keys."""
+        assert self._public_keys is not None, "The public_keys are not set."
+        return self._public_keys
+
+    @public_keys.setter
+    def public_keys(self, public_keys: Dict[str, str]) -> None:
+        """
+        Set the public keys.
+
+        :param public_keys: the public keys.
+        :return: None
+        """
+        self._public_keys = public_keys
+
+    @property
+    def addresses(self) -> Dict[str, str]:
+        """Get the addresses."""
+        assert self._addresses is not None, "The addresses are not set."
+        return self._addresses
+
+    @addresses.setter
+    def addresses(self, addresses: Dict[str, str]) -> None:
+        """
+        Set the addresses.
+
+        :param addresses: the addresses.
+        :return: None
+        """
+        self._addresses = addresses
+
+
 class Agent(ABC):
     """This class implements a template agent."""
 
     def __init__(
         self,
-        name: str,
+        identity: Identity,
```

Comment:
see above

Reply:
also, on another note: should the `Agent` have a crypto-identity?
If not (that is, `Agent` won't use the fields `address` and `public_keys`), then I propose two alternatives:

1. only `AEA` instances hold an `Identity` object. `Agent` still has a `name` in the constructor, but in case the `Agent` is actually the `AEA`, `Agent.name == AEA.identity.name`
2. Having a simpler identity for Agent, e.g. `AgentIdentity`, or simply `Identity`, which only holds a `name` (but later it can be extended). The Identity class for the `AEA` would be called `AEAIdentity`, which is a subclass of the previous identity, plus the addresses and public keys..

Reply:
The `Agent` should not have a crypto-identity as we suggest that it would be better to use an AEA for that. Of course, you could always extend our `Agent` base class to add a wallet.

Reply:
I don't like 1. The identity concept will be pretty quickly much broader than the `name`. The `name` is really just a logging helper, it is not used anywhere else, so we should not keep it as a first class argument. The `Agent` will need some `address` to connect to the OEF for instance. So (also see above) we might want to make `name` and `addresses` the required arguments in the identity constructor and drop `public_keys` as per suggestion by @Totoual . The subclass approach from 2. could be trivially added as we enrich the identity concept without breaking the AEA/Agent api

Reply:
agreed